daffodil-runtime1/src/main/scala/org/apache/daffodil/infoset/DataValue.scala

```diff
@@ -0,0 +1,209 @@
+  @inline def getBoolean = v.asInstanceOf[JBoolean]
+  @inline def getNumber = v.asInstanceOf[JNumber]
+  @inline def getInt = v.asInstanceOf[JInt]
+  @inline def getLong = v.asInstanceOf[JLong]
+  @inline def getDouble = v.asInstanceOf[JDouble]
+  @inline def getBigInt = v.asInstanceOf[JBigInt]
+  @inline def getString = v.asInstanceOf[JString]
+  @inline def getURI = v.asInstanceOf[URI]
+
+  @inline def getNonNullable = this.asInstanceOf[DataValue[T, X with NonNullable]]
+  @inline def getNullablePrimitive = this.asInstanceOf[DataValue.DataValuePrimitiveNullable]
+
+  @inline def getOptionAnyRef = {
+    if (isEmpty) {
+      None
+    } else {
+      Some(getAnyRef)
+    }
+  }
+
+  @inline def getMaybe[T <: AnyRef]: Maybe[T] = {
+    if (isEmpty) {
+      Nope
+    } else {
+      One(getAnyRef.asInstanceOf[T])
+    }
+  }
+}
+
+object DataValue {
+  type DataValueAny = DataValue[AnyRef, Nullable]
+  type DataValueNonEmpty = DataValue[AnyRef, NonNullable]
+  type DataValuePrimitive = DataValue[AnyRef, NonNullable with DataValuePrimitiveType]
+  type DataValuePrimitiveNullable = DataValue[AnyRef, Nullable with DataValuePrimitiveType]
+  type DataValuePrimitiveOrUseNilForDefaultOrNull = DataValue[AnyRef, tUseNilForDefault]
+
+  type DataValueEmpty = DataValue[Null, Nullable with DataValuePrimitiveType]
+  type DataValueBigDecimal = DataValue[JBigDecimal, NonNullable with DataValuePrimitiveType]
+  type DataValueCalendar = DataValue[DFDLCalendar, NonNullable with DataValuePrimitiveType]
+  type DataValueDateTime = DataValue[DFDLDateTime, NonNullable with DataValuePrimitiveType]
+  type DataValueDate = DataValue[DFDLDate, NonNullable with DataValuePrimitiveType]
+  type DataValueTime = DataValue[DFDLTime, NonNullable with DataValuePrimitiveType]
+  type DataValueByteArray = DataValue[Array[Byte], NonNullable with DataValuePrimitiveType]
+  type DataValueBool = DataValue[JBoolean, NonNullable with DataValuePrimitiveType]
+  type DataValueNumber = DataValue[JNumber, NonNullable with DataValuePrimitiveType]
+  type DataValueLong = DataValue[JLong, NonNullable with DataValuePrimitiveType]
+  type DataValueDouble = DataValue[JDouble, NonNullable with DataValuePrimitiveType]
+  type DataValueBigInt = DataValue[JBigInt, NonNullable with DataValuePrimitiveType]
+  type DataValueString = DataValue[JString, NonNullable with DataValuePrimitiveType]
+  type DataValueURI = DataValue[URI, NonNullable with DataValuePrimitiveType]
+  type DataValueFloat = DataValue[JFloat, NonNullable with DataValuePrimitiveType]
+  type DataValueByte = DataValue[JByte, NonNullable with DataValuePrimitiveType]
+  type DataValueInt = DataValue[JInt, NonNullable with DataValuePrimitiveType]
+  type DataValueShort = DataValue[JShort, NonNullable with DataValuePrimitiveType]
+  type DataValueDINode = DataValue[DINode, NonNullable with DataValuePrimitiveType]
+  type DataValueUseNilForDefault = DataValue[UseNilForDefaultObj, NonNullable]
```

Comment:
This one is the type for data values that can only be UseNilForDefault ? We don't need this do we?

Reply:
No. This is just copy pasta. Every type we use gets a type alias, an asFoo accessor, and an implicit toDataValue converter.

This seems simpler then deciding which types need which things.